### ISSUE TITLE
Understanding the code of Transformer Class

### ISSUE THREAD

**Issue by working12** [UKPLab/sentence-transformers#696]
In the `forward` function of the `Transformer`    
```
...
def forward(self, features):
        """Returns token_embeddings, cls_token"""
        output_states = self.auto_model(**features)
        output_tokens = output_states[0]

        cls_tokens = output_tokens[:, 0, :]  # CLS token is first token
        features.update({'token_embeddings': output_tokens, 'cls_token_embeddings': cls_tokens, 'attention_mask': features['attention_mask']})
...
```

Why are we storing the `cls_tokens` ? Because isn't the `[CLS]` token always embedded the same way by a model? So why are we doing it?

**Comment by nreimers** [comment]
Transformers are contexualized embeddings, hence, the embedding of the CLS token depends on all words of the sentence

**Comment by working12** [comment]
Thanks for the response.

I am asking all these because of the code in the Pooling layer. 

1. When we are saying that we are going for `pooling_mode_mean_tokens ` we are basically adding all these embedding of a sentence? (Each row denotes a sentence)
 
![image](https://user-images.githubusercontent.com/12879472/104812402-2fcbdb00-5802-11eb-8a56-125db6202967.png)

2. I also don't understand the use of attention mask, why are you using them? (Or any general reason why is it used in these kind of models?)

@nreimers

**Comment by nreimers** [comment]
1) Yes, we average the embeddings for all word pieces
2) Sentences can have different length. To encode them as a batch, we must bring them all to the same length by adding padding tokens. In order that these padding tokens do not change the computed average, we use the attention mask to keep track which tokens are real tokens and which are padding tokens. So we ignore padding tokens at the computation of the sentence embedding

**Comment by working12** [comment]
@nreimers  Thanks for the response. But if I add all those blue vectors then the encoding for a sentence would be `(1, number of token in sentence)`. Then two different sentence can have two different vector length? How can we calculate cosine distance between them?

The blue ones represent embedding of a sentence. And then if we add them, we get average embedding. You get why I am getting confused?

**Comment by nreimers** [comment]
Every word piece is encoded to a 768 dimensional vector by BERT. Then, we have the mean pooling operations that computes the mean of all vectors. This gives another 768 dimensional vector, independent how many tokens / word pieces are in your sentence

**Comment by working12** [comment]
So that means ultimately a sentence becomes a 768 dimensional vector in the end irrespective of how many tokens it has, because well we are doing the mean, right? (I feel I understand it correctly )

**Comment by nreimers** [comment]
Correct

**Comment by working12** [comment]
Thanks for confirming @nreimers . You are great as always.
